usr/lib/cca_stdll/cca_specific.c

```diff
@@ -1090,6 +1092,231 @@ CK_RV token_specific_rsa_decrypt(STDLL_TokData_t * tokdata,
+    }
+
+    /* The max value allowable by CCA for out_data_len is 512, so cap the
+     * incoming value if its too large. CCA will throw error 8, 72 otherwise.
+     */
+    if (*out_data_len > 512)
+        *out_data_len = 512;
+
+    rule_array_count = 2;
+    switch (oaep->hashAlg) {
+    case CKM_SHA_1:
+        if (oaep->mgf != CKG_MGF1_SHA1) {
+            TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+            rc = CKR_MECHANISM_PARAM_INVALID;
+            goto done;
+        }
+        memcpy(rule_array, "PKCSOAEPSHA-1   ", 2 * CCA_KEYWORD_SIZE);
+        break;
+
+    case CKM_SHA256:
+        if (oaep->mgf != CKG_MGF1_SHA256) {
+            TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+            rc = CKR_MECHANISM_PARAM_INVALID;
+            goto done;
+        }
+        memcpy(rule_array, "PKCSOAEPSHA-256 ", 2 * CCA_KEYWORD_SIZE);
+        break;
+
+    default:
+        TRACE_ERROR("%s\n", ock_err(ERR_MECHANISM_PARAM_INVALID));
+        rc = CKR_MECHANISM_PARAM_INVALID;
+        goto done;
+    }
+
+    data_structure_length = 0;
+
+    dll_CSNDPKE(&return_code,
+                &reason_code,
+                NULL, NULL,
+                &rule_array_count,
+                rule_array,
+                (long *)&in_data_len,
+                in_data,
+                &data_structure_length,  // must be 0
+                NULL,           // ignored
+                (long *)&(attr->ulValueLen),
+                attr->pValue,
+                (long *)out_data_len,
+                out_data);
+
+    if (return_code != CCA_SUCCESS) {
+        TRACE_ERROR("CSNDPKE (RSA ENCRYPT) failed. return:%ld, reason:%ld\n",
+                    return_code, reason_code);
+        rc = CKR_FUNCTION_FAILED;
+        goto done;
+    } else if (reason_code != 0) {
+        TRACE_WARNING("CSNDPKE (RSA ENCRYPT) succeeded, but"
+                      " returned reason:%ld\n", reason_code);
+    }
+
+done:
+    object_put(tokdata, key_obj, TRUE);
+    key_obj = NULL;
+
+    return rc;
+}
+
+CK_RV token_specific_rsa_oaep_decrypt(STDLL_TokData_t *tokdata,
+                                      ENCR_DECR_CONTEXT *ctx,
+                                      CK_BYTE *in_data,
+                                      CK_ULONG in_data_len,
+                                      CK_BYTE *out_data,
+                                      CK_ULONG *out_data_len,
+                                      CK_BYTE *hash,
+                                      CK_ULONG hlen)
+{
+    CK_RSA_PKCS_OAEP_PARAMS *oaep;
+    long return_code, reason_code, rule_array_count, data_structure_length;
+    unsigned char rule_array[CCA_RULE_ARRAY_SIZE] = { 0, };
+    CK_ATTRIBUTE *attr;
+    OBJECT *key_obj = NULL;
+    CK_RV rc;
+
+    UNUSED(tokdata);
+    UNUSED(hash);
+    UNUSED(hlen);
+
+    rc = object_mgr_find_in_map1(tokdata, ctx->key, &key_obj, READ_LOCK);
+    if (rc != CKR_OK) {
+        TRACE_DEVEL("object_mgr_find_in_map1 failed\n");
```

Comment:
return rc instead of goto to avoid the object_put() ?

Reply:
see above